### DISCUSSION TITLE
Feature request: Add side_data attribute to Stream

### DISCUSSION THREAD

**Discussion by hyenal** [PyAV-Org/PyAV#1629] [2. Feature Ideas]
## Overview

Sometimes a stream encodes side data in the entire stream (in [side_data](https://ffmpeg.org/doxygen/trunk/structAVStream.html)). Some of those side data are quite essential to read the video correctly.
For instance `displaymatrix` sometimes encode whether the video is rotated or not.

From the AV documentation it seems like we could know easily if there is side data with the attribute `nb_side_data`.

I would like to know if this is an interesting feature for this library ? I started a fork [there](https://github.com/hyenal/PyAV/tree/add-stream-side-data) with, for now, a simple skeleton of how it would look like

## Desired Behavior

When reading a video (or any other stream) you would get the side data as a dictionary:
```python
# This would give you the stream metadata
container.stream.video[0].side_data
```

**Comment by hyenal** [comment]
I made [this PR](https://github.com/hyenal/PyAV/pull/1/files) on my own fork. It only fetch the DISPLAYMATRIX but would be happy to implement more and submit a PR to this repo if anyone is interested

**Comment by lgeiger** [comment]
@hyenal I'd be really interested in having this functionality upstreamed to `av` as it would fix #1045 is a blocker for me currently and also many people seem to run into.

@jlaine or @WyattBlue are probably in a better position to judge how this should be implemented best to make it consistent with the way frame side data is handled. I'm happy to help out and write some test cases if that would be helpful.

**Comment by hyenal** [comment]
Thanks @lgeiger, so happy to see this getting momentum!

It looks like it's been merged in @WyattBlue repository and will be available in PyAv soon? 

I am also happy to write some tests if this helps :)

**Comment by lgeiger** [comment]
> It looks like it's been merged in @WyattBlue repository and will be available in PyAv soon?

Yes, it's already available on [pypi](https://pypi.org/project/pyav) though this causes dependency conflicts for me as I am relying on a different package that requires `av` directly which is why it would be great to get upstreamed at some point.

**Comment by hyenal** [comment]
> Yes, it's already available on [pypi](https://pypi.org/project/pyav) though this causes dependency conflicts for me as I am relying on a different package that requires `av` directly which is why it would be great to get upstreamed at some point.

I see thanks for the explanation!

Let's wait for an answer from maintainers then, I am hopeful this could be merged 🙏

**Comment by WyattBlue** [comment]
Sure, @hyenal Make a new PR for this repo.

**Comment by hyenal** [comment]
> I'm happy to help out and write some test cases if that would be helpful.

@lgeiger do you have some tests in mind ? As far as I understood we rely on FATE to get video samples, I hope those have a sample with a non zero display matrix so that we could write a test easily.
Otherwise the remaining work I had in mind was to make sure stream.sidedata works and polish the code (removing the two TODOs)

I intend to do this next week so if you it's a blocker for you I d be happy to invite you as a contributor to my PR :)

**Comment by lgeiger** [comment]
> @lgeiger do you have some tests in mind ? As far as I understood we rely on FATE to get video samples, I hope those have a sample with a non zero display matrix so that we could write a test easily.

Not really. I probably could share some example videos but I guess it would be better to get them from FATE.

> I intend to do this next week so if you it's a blocker for you I d be happy to invite you as a contributor to my PR :)

That's great, thanks for working on it. It's not a blocker for me anymore. For reference, we ended up using [pymediainfo](https://pymediainfo.readthedocs.io/en/stable/) to read the rotation information from the video files and handle them accordingly via Numpy which works well for our use case but it would still be great if this would be natively supported in av.

**Comment by WyattBlue** [comment]
We're planning to remove Stream.side_data (https://github.com/PyAV-Org/PyAV/pull/1570) in PyAV 14 since FFmpeg doesn't even document `AVStream.side_data` or `AVStream.nb_side_data` anymore.

I think there will be some alternative API, but we're waiting for FFmpeg 8 to be released.

**Comment by lgeiger** [comment]
@WyattBlue thanks for the heads up. What's the recommended way of checking whether the video content needs to be rotated with PyAV?

As far as I can tell without `side_data` we won't know whether the frame needs to be rotated (e.g. when we need to do further processing of the numpy frame).

Of course this can be checked via `ffprobe -loglevel error -select_streams v:0 -show_entries stream_tags=rotate -of default=nw=1:nk=1 video.mp4` or using something like [pymediainfo](https://pymediainfo.readthedocs.io/en/stable/) but that means adding another requirement which tends to complicate things.

As far as I can tell correct handling of rotation is the reason why some upstream packages still rely on av < 10 (see #1045).

**Reply by burningion** [reply]
Ran into this today, have an updated version of pyav, where the side_data went away. It's unfortunate, but the world's most popular camera uses this when recording video.

**Comment by hmaarrfk** [comment]
It would be great if we can "write" this side data ourselves as well.

Looking at the source:
https://ffmpeg.org/doxygen/7.0/android__camera_8c_source.html#l00640

1. Create new side packet `AV_PKT_DATA_DISPLAYMATRIX`.
2. Fill it with the right value of a matrix.

It would be great if we can replicate this.
